spec/core/ics-003-connection-semantics/UPGRADES.md

```diff
@@ -0,0 +1,407 @@
+
+    // verify proofs of counterparty state
+    abortTransactionUnless(currentConnection.client.verifyConnectionState(proofHeight, proofConnection, counterpartyConnection))
+    abortTransactionUnless(currentConnection.client.verifyUpgradeStatus(proofHeight, proofUpgradeStatus, counterpartyUpgradeStatus))
+
+    // counterparty must be in TRY state
+    restoreConnectionUnless(counterpartyStatus.state == UPGRADE_TRY)
+    restoreConnectionUnless(counterpartyConnection.State == UPGRADE_TRY)
+
+    // verify connections are mutually compatible
+    // this will also check counterparty chosen version is valid
+    restoreConnectionUnless(IsCompatible(counterpartyConnection, connection))
+
+    // upgrade is complete
+    // set connection to OPEN and remove unnecessary state
+    currentConnection.state = OPEN
+    provableStore.set(connectionPath(identifier), currentConnection)
+    provableStore.delete(statusPath(identifier))
+    privateStore.delete(restorePath(identifier))
+}
+```
+
+```typescript
+function onChanUpgradeConfirm(
+    identifier: Identifier,
+    counterpartyConnection: ConnectionEnd,
+    proofConnection: CommitmentProof,
+    proofHeight: Height,
+) {
+    // current connection is in UPGRADE_TRY
+    currentConnection = provableStore.get(connectionPath(identifier))
+    abortTransactionUnless(connection.state == UPGRADE_TRY)
+
+    // counterparty must be in OPEN state
+    abortTransactionUnless(counterpartyConnection.State == OPEN)
+
+    // verify proofs of counterparty state
+    abortTransactionUnless(currentConnection.client.verifyConnectionState(proofHeight, proofConnection, counterpartyConnection))
+    
+    // upgrade is complete
+    // set connection to OPEN and remove unnecessary state
+    currentConnection.state = OPEN
+    provableStore.set(connectionPath(identifier), currentConnection)
+    provableStore.delete(statusPath(identifier))
+    privateStore.delete(restorePath(identifier))
+}
+```
+
+Note: In the `TRY` and `ACK` steps, the chain has the ability to abort the upgrade process if the upgrade parameters are not suitable or the timeout has exceeded. In this case, the chain is expected to write `UPGRADE_ERR` as the state in the `UPGRADE_STATUS`, and restore the original connection with state `OPEN`. The counterparty can then verify that the upgrade status is `UPGRADE_ERR` and restore its original connection to `OPEN` as well, thus cancelling the upgrade.
+
+```typescript
+function restoreConnectionUnless(condition: bool) {
+    if !condition {
+        // cancel upgrade
+        // set UpgradeStatus to `UPGRADE_ERR`
+        // and restore original conneciton
+        errStatus = UpgradeStatus{UPGRADE_ERR}
+        provableStore.set(statusPath(identifier), errStatus)
+        originalConnection = privateStore.get(restorePath(identifier))
+        provableStore.set(connectionPath(identifier), originalConnection)
+        privateStore.delete(restorePath(identifier))
+        // caller should return as well
```

Comment:
This is to continue with the pattern of `abortTransactionUnless`, and this is semantically similar since we are aborting the handshake. So i think it makes sense to follow the same pattern